### PR TITLE
Resume videos for non-logged in users

### LINKED ISSUE #3866: Resume video for non-logged in users
**Describe the problem to be solved**
As discussed in #660 it's a nice feature to be able to resume a video previously started. On mobile phones it can be especially convenient since it's more common with a bad internet connection or other disturbance. The problem is that this is only working for logged in users. I'm not sure about how the user behavior look like, but personally I'm not often logged in to sites on my mobile device.

**Describe the solution you would like:**
As non-logged in user I'd like the videos to be resumed when I revisit/reload the watch page.

**Describe alternatives you have considered**  
Store the current duration in local storage. The question is how to handle the duration when a non-logged in user logs in and there's already a duration stored at the server side? The easiest may be to just remove the duration from local storage upon login.

### PR DESCRIPTION
**Pull request by kontrollanten**
## Description
Stores the current video duration in local storage for non-logged in users so that they can go back to a video and continue where they ended.

Upon app init a check is done to cleanup old durations.

## Related issues
closes #3866

## Has this been tested?

- [x] 👍 yes, light tests as follows are enough

## Resume
1) Logout.
1) Watch a video for a few seconds.
1) Reload the page.

## Cleanup
1) Edit the local storage and change the date in `peertube-videojs-video-watch` to a date more than 31 days back.
1) Reload the page.
1) See that the duration has been cleared.

### CODE REVIEW

File: client/src/app/app.component.ts [hunk 1]
```diff
@@ -99,6 +100,8 @@ export class AppComponent implements OnInit, AfterViewInit {
     this.openModalsIfNeeded()
 
     this.document.documentElement.lang = getShortLocale(this.localeId)
+
+    setTimeout(cleanupVideoWatch, 5)
```
Comment (kontrollanten) [inline client/src/app/app.component.ts:104]:
```suggestion
    setTimeout(cleanupVideoWatch, 1500) // Run in timeout to ensure we're not blocking the UI
```

File: client/src/assets/player/peertube-player-local-storage.ts [hunk 2]
```diff
@@ -68,6 +68,54 @@ function getStoredLastSubtitle () {
...
+  }))
+}
+
+function getVideoWatch(videoUUID?: string) {
+  let data
+
+  try {
+    data = JSON.parse(getLocalStorage('video-watch'))
+  } catch (error) {
+    console.error(error)
```
Comment (Chocobozzz) [inline client/src/assets/player/peertube-player-local-storage.ts:90]: "Cannot parse video watch history from local storage"

File: client/src/assets/player/peertube-player-local-storage.ts [hunk 3]
```diff
@@ -68,6 +68,54 @@ function getStoredLastSubtitle () {
...
+
+    if (diff > 30) return acc
+
+    return {
+      ...acc,
+      [videoUUID]: data[videoUUID]
+    }
+  }, {})
+
+  setLocalStorage('video-watch', JSON.stringify(newData))
```
Comment (Chocobozzz) [inline client/src/assets/player/peertube-player-local-storage.ts:116]: `video-watch-history`

File: client/src/assets/player/peertube-player-local-storage.ts [hunk 4]
```diff
@@ -68,6 +68,54 @@ function getStoredLastSubtitle () {
   return getLocalStorage('last-subtitle')
 }
 
+function saveVideoWatch(videoUUID: string, duration: number) {
+  const data = getVideoWatch()
+  const now = new Date()
+
+  return setLocalStorage(`video-watch`, JSON.stringify({
```
Comment (Chocobozzz) [inline client/src/assets/player/peertube-player-local-storage.ts:75]: `video-watch-history`

File: client/src/app/app.component.ts [hunk 5]
```diff
@@ -99,6 +100,8 @@ export class AppComponent implements OnInit, AfterViewInit {
     this.openModalsIfNeeded()
 
     this.document.documentElement.lang = getShortLocale(this.localeId)
+
+    setTimeout(cleanupVideoWatch, 1500) // Run in timeout to ensure we're not blocking the UI
```
Comment (Chocobozzz) [inline client/src/app/app.component.ts:104]: Move this in video-watch component to not import player stuff in the main component

File: client/src/assets/player/peertube-player-local-storage.ts [hunk 6]
```diff
@@ -68,6 +68,54 @@ function getStoredLastSubtitle () {
...
+    ...data,
+    [videoUUID]: {
+      duration,
+      date: `${now.getFullYear()}-${now.getUTCMonth() + 1}-${now.getUTCDate()}`
```
Comment (Chocobozzz) [inline client/src/assets/player/peertube-player-local-storage.ts:79]: Why don't you use `.toISOString()`?

File: client/src/assets/player/peertube-plugin.ts [hunk 7]
```diff
@@ -187,8 +188,12 @@ class PeerTubePlugin extends Plugin {
       if (currentTime - lastCurrentTime >= 1) {
         lastCurrentTime = currentTime
 
-        this.notifyUserIsWatching(currentTime, options.url, options.authorizationHeader)
+        if (options) {
+          this.notifyUserIsWatching(currentTime, options.url, options.authorizationHeader)
           .catch(err => console.error('Cannot notify user is watching.', err))
```
Comment (Chocobozzz) [inline client/src/assets/player/peertube-plugin.ts:193]: Please indent this catch

**Review by Chocobozzz — CHANGES_REQUESTED** [review]

**Comment by Chocobozzz** [comment]
Thanks

**Pull request merged**